src/clients/http.rs

```diff
@@ -15,41 +15,175 @@
+    }
+}
+
+impl Deref for Response {
+    type Target = hyper::http::response::Response<BoxBody<Bytes, hyper::Error>>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl From<hyper::http::response::Response<Incoming>> for Response {
+    fn from(response: hyper::http::Response<Incoming>) -> Self {
+        Response(response.map(|body| body.boxed()))
+    }
+}
+
+pub type HttpClientInner =
+    hyper_util::client::legacy::Client<HttpsConnector<HttpConnector>, BoxBody<Bytes, hyper::Error>>;
+
+/// A trait implemented by all clients that use HTTP for their inner client.
+pub trait HttpClientExt: Client {
+    fn inner(&self) -> &HttpClient;
+}
 
+/// An HTTP client wrapping an inner `hyper` HTTP client providing a higher-level API.
 #[derive(Clone)]
 pub struct HttpClient {
     base_url: Url,
     health_url: Url,
-    client: reqwest::Client,
+    inner: HttpClientInner,
 }
 
 impl HttpClient {
-    pub fn new(base_url: Url, client: reqwest::Client) -> Self {
+    pub fn new(base_url: Url, inner: HttpClientInner) -> Self {
         let health_url = base_url.join("health").unwrap();
         Self {
             base_url,
             health_url,
-            client,
+            inner,
         }
     }
 
     pub fn base_url(&self) -> &Url {
         &self.base_url
     }
 
+    pub fn endpoint(&self, path: impl Into<&'static str>) -> Url {
+        self.base_url.join(path.into()).unwrap()
+    }
+
+    #[instrument(skip_all, fields(url))]
+    pub async fn get(
+        &self,
+        url: Url,
+        headers: HeaderMap,
+        body: impl Serialize,
+    ) -> Result<Response, Error> {
+        self.send(url, Method::GET, headers, body).await
+    }
+
+    #[instrument(skip_all, fields(url))]
+    pub async fn post(
+        &self,
+        url: Url,
+        headers: HeaderMap,
+        body: impl Serialize,
+    ) -> Result<Response, Error> {
+        self.send(url, Method::POST, headers, body).await
+    }
+
+    #[instrument(skip_all, fields(url))]
+    pub async fn send(
+        &self,
+        url: Url,
+        method: Method,
+        headers: HeaderMap,
+        body: impl Serialize,
+    ) -> Result<Response, Error> {
+        let ctx = Span::current().context();
+        let headers = trace::with_traceparent_header(&ctx, headers);
+        let mut builder = hyper::http::request::Builder::new()
+            .method(method)
+            .uri(url.as_uri());
+        match builder.headers_mut() {
+            Some(headers_mut) => {
+                headers_mut.extend(headers.clone());
+                debug!(
+                    ?url,
+                    ?headers,
+                    // If the body can not be serialized it will be caught below, we don't want to
+                    // complicate logging by handling string serialization error here.
+                    body = serde_json::to_string(&body).unwrap_or("".to_string()),
+                    "sending client request"
+                );
```

Comment:
Move `headers_mut.extend()` after `debug!()` and drop the unnecessary clone. 

Hmm, I don't really like the `serde_json::to_string()` embedded here. Since all of our types impl `Debug`, we can alternatively update `body: impl Serialize + std::fmt::Debug` throughout and simplify this to:
```rust
debug!(?url, ?headers, ?body, "sending client request");
```

Reply:
> alternatively update body: impl Serialize + std::fmt::Debug throughout

Going back to the `RequestLike` empty trait from before: In this case what is your opinion on having a trait to blanket `Serialize + Debug` so that we can simplify it to `impl RequestLike` or `impl IntoRequest` or something instead of making every function signature lengthier (also just like the idea that if I ever need to know what contrains a request or need to make modifications to it, its a single source of truth rather than several function signature changes)

(also addressed other nits and will push changes momentarily)

Reply:
Yea, I agree it makes it easier to have a single source of truth if we need to change it later.

Reply:
Changes on this look good, thanks!